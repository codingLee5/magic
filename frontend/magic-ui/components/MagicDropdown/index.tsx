import type { DropDownProps } from "antd"
import { Dropdown } from "antd"
import { cx } from "antd-style"
import { useStyles } from "./style"

export type MagicDropdownProps = DropDownProps

function MagicDropdown({
	menu: { rootClassName, ...menu } = {},
	overlayClassName,
	...props
}: MagicDropdownProps) {
	const { styles } = useStyles()

	return (
		<Dropdown
			overlayClassName={cx(styles.dropdown, overlayClassName)}
			menu={{
				rootClassName: cx(styles.dropdown, styles.subMenu, rootClassName),
				...menu,
			}}
			{...props}
		/>
	)
}

export default MagicDropdown
