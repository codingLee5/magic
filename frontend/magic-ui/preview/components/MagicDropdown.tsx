import React from "react"
import { Space } from "antd"
import MagicDropdown from "../../components/MagicDropdown"
import MagicButton from "../../components/MagicButton"
import ComponentDemo from "./Container"

const DropdownDemo: React.FC = () => {
	const menuItems = [
		{ key: "1", label: "菜单项1" },
		{ key: "2", label: "菜单项2" },
		{ key: "3", label: "菜单项3" },
	]

	const nestedMenuItems = [
		{ key: "1", label: "菜单项1" },
		{
			key: "2",
			label: "子菜单",
			children: [
				{ key: "2-1", label: "子菜单项1" },
				{ key: "2-2", label: "子菜单项2" },
			],
		},
		{ key: "3", label: "菜单项3" },
	]

	return (
		<div>
			<ComponentDemo
				title="基础下拉菜单"
				description="最基本的下拉菜单组件"
				code="<MagicDropdown menu={{ items }}><Button>下拉菜单</Button></MagicDropdown>"
			>
				<Space>
					<MagicDropdown
						menu={{
							items: menuItems,
						}}
					>
						<MagicButton>下拉菜单</MagicButton>
					</MagicDropdown>
				</Space>
			</ComponentDemo>

			<ComponentDemo title="嵌套菜单" description="支持嵌套子菜单" code="children">
				<Space>
					<MagicDropdown
						menu={{
							items: nestedMenuItems,
						}}
					>
						<MagicButton>嵌套菜单</MagicButton>
					</MagicDropdown>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="不同触发方式"
				description="支持不同的触发方式"
				code="trigger: 'click' | 'hover' | 'contextMenu'"
			>
				<Space>
					<MagicDropdown
						trigger={["click"]}
						menu={{
							items: menuItems,
						}}
					>
						<MagicButton>点击触发</MagicButton>
					</MagicDropdown>
					<MagicDropdown
						trigger={["hover"]}
						menu={{
							items: menuItems,
						}}
					>
						<MagicButton>悬停触发</MagicButton>
					</MagicDropdown>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="不同位置"
				description="支持不同的显示位置"
				code="placement: 'top' | 'bottom' | 'left' | 'right'"
			>
				<Space>
					<MagicDropdown
						placement="top"
						menu={{
							items: menuItems,
						}}
					>
						<MagicButton>顶部显示</MagicButton>
					</MagicDropdown>
					<MagicDropdown
						placement="bottom"
						menu={{
							items: menuItems,
						}}
					>
						<MagicButton>底部显示</MagicButton>
					</MagicDropdown>
				</Space>
			</ComponentDemo>

			<ComponentDemo title="事件处理" description="监听菜单点击事件" code="onClick">
				<Space>
					<MagicDropdown
						menu={{
							items: menuItems,
							onClick: ({ key }) => {
								console.log("点击的菜单项:", key)
							},
						}}
					>
						<MagicButton>事件处理</MagicButton>
					</MagicDropdown>
				</Space>
			</ComponentDemo>
		</div>
	)
}

export default DropdownDemo
