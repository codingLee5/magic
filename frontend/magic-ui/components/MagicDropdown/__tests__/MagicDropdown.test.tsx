import { render } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicDropdown from "../index"

const renderWithTheme = (component: React.ReactElement) =>
	render(<MagicThemeProvider theme="light">{component}</MagicThemeProvider>)

describe("MagicDropdown", () => {
	it("应该正常渲染", () => {
		renderWithTheme(<MagicDropdown menu={{ items: [] }}>下拉</MagicDropdown>)
		// 只要不报错即可
		expect(true).toBe(true)
	})

	// 快照测试
	describe("快照测试", () => {
		it("基础下拉菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicDropdown menu={{ items: [] }}>下拉菜单</MagicDropdown>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带菜单项下拉菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicDropdown
					menu={{
						items: [
							{ key: "1", label: "菜单项1" },
							{ key: "2", label: "菜单项2" },
						],
					}}
				>
					下拉菜单
				</MagicDropdown>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带图标下拉菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicDropdown
					menu={{
						items: [
							{ key: "1", label: "菜单项1", icon: <span>🏠</span> },
							{ key: "2", label: "菜单项2", icon: <span>⚙️</span> },
						],
					}}
				>
					<div>🏠 下拉菜单</div>
				</MagicDropdown>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("禁用状态下拉菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicDropdown disabled menu={{ items: [] }}>
					下拉菜单
				</MagicDropdown>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带自定义属性下拉菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicDropdown data-testid="custom-dropdown" menu={{ items: [] }}>
					下拉菜单
				</MagicDropdown>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带类名下拉菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicDropdown className="custom-dropdown" menu={{ items: [] }}>
					下拉菜单
				</MagicDropdown>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带按钮样式下拉菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicDropdown menu={{ items: [] }}>
					<button>按钮下拉</button>
				</MagicDropdown>,
			)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
