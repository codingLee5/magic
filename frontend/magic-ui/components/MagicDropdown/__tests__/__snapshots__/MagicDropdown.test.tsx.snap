// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicDropdown > 快照测试 > 基础下拉菜单快照 1`] = `
<DocumentFragment>
  <span
    class="magic-dropdown-trigger"
  >
    下拉菜单
  </span>
</DocumentFragment>
`;

exports[`MagicDropdown > 快照测试 > 带图标下拉菜单快照 1`] = `
<DocumentFragment>
  <div
    class="magic-dropdown-trigger"
  >
    🏠 下拉菜单
  </div>
</DocumentFragment>
`;

exports[`MagicDropdown > 快照测试 > 带按钮样式下拉菜单快照 1`] = `
<DocumentFragment>
  <button
    class="magic-dropdown-trigger"
  >
    按钮下拉
  </button>
</DocumentFragment>
`;

exports[`MagicDropdown > 快照测试 > 带类名下拉菜单快照 1`] = `
<DocumentFragment>
  <span
    class="magic-dropdown-trigger custom-dropdown"
  >
    下拉菜单
  </span>
</DocumentFragment>
`;

exports[`MagicDropdown > 快照测试 > 带自定义属性下拉菜单快照 1`] = `
<DocumentFragment>
  <span
    class="magic-dropdown-trigger"
  >
    下拉菜单
  </span>
</DocumentFragment>
`;

exports[`MagicDropdown > 快照测试 > 带菜单项下拉菜单快照 1`] = `
<DocumentFragment>
  <span
    class="magic-dropdown-trigger"
  >
    下拉菜单
  </span>
</DocumentFragment>
`;

exports[`MagicDropdown > 快照测试 > 禁用状态下拉菜单快照 1`] = `
<DocumentFragment>
  <span
    class="magic-dropdown-trigger"
    disabled=""
  >
    下拉菜单
  </span>
</DocumentFragment>
`;
